File: solutions/python/arrays-hashing/longest-palindrome.py
# LeetCode 409 — Longest Palindrome (Easy)
# Category: Arrays & Hashing · Approach: Counting
# Time: O(n) | Space: O(1)
# Source: https://leetcode.com/problems/longest-palindrome/

from collections import Counter


def longest_palindrome(s: str) -> int:
    counts = Counter(s)
    length = 0
    has_odd = False
    for c in counts.values():
        length += (c // 2) * 2
        if c % 2 == 1:
            has_odd = True
    return length + (1 if has_odd else 0)

File: solutions/typescript/arrays-hashing/longest-palindrome.ts
// LeetCode 409 — Longest Palindrome (Easy)
// Category: Arrays & Hashing · Approach: Counting
// Time: O(n) | Space: O(1)
// Source: https://leetcode.com/problems/longest-palindrome/

function longestPalindrome(s: string): number {
  const counts = new Map<string, number>();
  for (const ch of s) counts.set(ch, (counts.get(ch) ?? 0) + 1);
  let length = 0;
  let hasOdd = false;
  for (const c of counts.values()) {
    length += Math.floor(c / 2) * 2;
    if (c % 2 === 1) hasOdd = true;
  }
  return length + (hasOdd ? 1 : 0);
}
